### PR TITLE
S-Expression lexer primitives and predicates

### PR DESCRIPTION
**Pull request by tealeg**
This PR puts in place the first building block of the lexer - a series of Token types and predicates that identify the token about to be scanned.  

Note that we already see a potential "special case" in our language - the `-` character can start a number or be an symbol referencing an operator.  Lisp implimentations solve this issue in varied ways, most rely on knowing that function references look different (i.e Common Lisp would referencet the `-` function as `#'-`) and that implicit invocations only occur on symbols that appear at the head of a list-form s-expression.  Some go as far as doing in-line expansions in the reader to convert negated numbers into the logical form (i.e. converting `-1` to `(- 0 1)`).    We'll probably just force the token to the correct type in the Parser instead. 

Fixes: #14 
Fixes: #15

### CODE REVIEW

File: sexpr/lexer.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	NUMBER
+	BOOL
+	COMMENT
+	SYMBOL
+)
+
+// isWhitespace returns true if the rune is the first rune of a
+// Whitespace sequence.
+func isWhitespace(r rune) bool {
+	return r == ' ' || r == '\t' || r == '\r' || r == '\n'
```
Comment (asdine) [inline sexpr/lexer.go]: :tulip: The unicode package provides a [`IsSpace` function](https://golang.org/pkg/unicode/#IsSpace) that deals with additional special characters. Do you think we should use it?

File: sexpr/lexer.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package sexpr
```
Comment (asdine) [inline sexpr/lexer.go]: The `sexpr` package is directly related to the `rule` package, what about moving it under `rule/sexpr` ?

File: sexpr/lexer_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	require.False(t, isLParen('\n'))
+	require.False(t, isLParen('-'))
+	require.False(t, isLParen('a'))
+	require.False(t, isLParen('"'))
+	require.False(t, isLParen(')'))
+	require.False(t, isLParen('_'))
+	require.False(t, isLParen('0'))
+	require.False(t, isLParen('#'))
+	require.False(t, isLParen(';'))
+}
```
Comment (YassineE) [inline sexpr/lexer_internal_test.go]: 🌷🌷 maybe add: `require.False(t, isLParen(')')`.

File: sexpr/lexer.go (hunk 1 above)
Comment (tealeg) [reply sexpr/lexer.go]: Yes!  This was buzzing in my head on Friday - I knew I'd seen it somewhere.

File: sexpr/lexer.go (hunk 2 above)
Comment (tealeg) [reply sexpr/lexer.go]: Good idea.

File: sexpr/lexer_internal_test.go (hunk 3 above)
Comment (YassineE) [reply sexpr/lexer_internal_test.go]: My bad 🙈

**Pull request merged**